src/System.Private.CoreLib/shared/System/IO/FileStream.Windows.cs

```diff
@@ -62,7 +62,14 @@ private void Init(FileMode mode, FileShare share)
             if (fileType != Interop.Kernel32.FileTypes.FILE_TYPE_DISK)
             {
                 _fileHandle.Dispose();
-                throw new NotSupportedException(SR.NotSupported_FileStreamOnNonFiles);
+
+                var lastWin32Error = Marshal.GetLastWin32Error();
```

Comment:
I don't think this does what you want it to do (did you test it?).  _fileHandle.Dispose P/Invokes with SetLastError = true, so it's going to overwrite the last error code from the call to GetFileType.